accounts/abi/bind/dep_tree.go

```diff
@@ -0,0 +1,234 @@
+	input         map[string][]byte // map of the root contract pattern to the constructor input (if there is any)
+	deploy        DeployFn
+}
+
+// linkAndDeploy recursively deploys a contract and its dependencies:  starting by linking/deploying its dependencies.
+// The deployment result (deploy addresses/txs or an error) is stored in the depTreeDeployer object.
+func (d *depTreeDeployer) linkAndDeploy(node *depTreeNode) error {
+	// don't deploy contracts specified as overrides.  don't deploy their dependencies.
+	if node.overrideAddr != nil {
+		return nil
+	}
+
+	// if this contract/library depends on other libraries deploy them (and their dependencies) first
+	for _, childNode := range node.children {
+		if err := d.linkAndDeploy(childNode); err != nil {
+			return err
+		}
+	}
+	// if we just deployed any prerequisite contracts, link their deployed addresses into the bytecode to produce
+	// a deployer bytecode for this contract.
+	deployerCode := node.unlinkedCode
+	for _, child := range node.children {
+		linkAddr := d.deployedAddrs[child.pattern]
+		if child.overrideAddr != nil {
+			linkAddr = *child.overrideAddr
+		}
+		deployerCode = strings.ReplaceAll(deployerCode, "__$"+child.pattern+"$__", strings.ToLower(linkAddr.String()[2:]))
+	}
+
+	// Finally, deploy the contract.
+	addr, tx, err := d.deploy(d.input[node.pattern], common.Hex2Bytes(deployerCode))
+	if err != nil {
+		return err
+	}
+
+	d.deployedAddrs[node.pattern] = addr
+	d.deployerTxs[node.pattern] = tx
+	return nil
+}
+
+// result returns a result for this deployment, or an error if it failed.
+func (d *depTreeDeployer) result() *DeploymentResult {
+	return &DeploymentResult{
+		Txs:   d.deployerTxs,
+		Addrs: d.deployedAddrs,
+	}
+}
+
+func newDepTreeDeployer(deploy DeployFn) *depTreeDeployer {
+	return &depTreeDeployer{
+		deploy:        deploy,
+		deployedAddrs: make(map[string]common.Address),
+		deployerTxs:   make(map[string]*types.Transaction)}
+}
+
+// LinkAndDeploy deploys a specified set of contracts and their dependent
+// libraries.  If an error occurs, only contracts which were successfully
+// deployed are returned in the result.
+func LinkAndDeploy(deployParams *DeploymentParams, deploy DeployFn) (res *DeploymentResult, err error) {
+	unlinkedContracts := make(map[string]string)
+	accumRes := &DeploymentResult{
+		Txs:   make(map[string]*types.Transaction),
+		Addrs: make(map[string]common.Address),
+	}
+	for _, meta := range deployParams.contracts {
+		unlinkedContracts[meta.Pattern] = meta.Bin[2:]
+	}
+	treeBuilder := newDepTreeBuilder(deployParams.overrides, unlinkedContracts)
+	deps, _ := treeBuilder.BuildDepTrees()
+
+	for _, tr := range deps {
+		deployer := newDepTreeDeployer(deploy)
+		if deployParams.inputs != nil {
+			deployer.input = map[string][]byte{tr.pattern: deployParams.inputs[tr.pattern]}
+		}
+		err := deployer.linkAndDeploy(tr)
+		res := deployer.result()
+		accumRes.Accumulate(res)
+		if err != nil {
+			return accumRes, err
+		}
```

Comment:
```suggestion
		if err := deployer.linkAndDeploy(tr); err != nil {
			return accumRes, err
		}
		accumRes.Accumulate(deployer.result())
```
Afaict, if it `linkAndDeploy` exits with errors, there's nothing new to add to the accumulated result, so might aswell write it a bit more go-idiomatic

Reply:
If the call to `linkAndDeploy` deploys multiple contracts and then fails, we want the accumulated result to include whatever succeeded.

Reply:
Right. I didn't see that there's a recursive `linkAndDeploy` internally, which might add things _and_ still return with error later.